task6/js/script.js

```diff
@@ -0,0 +1,521 @@
+    let adsToShow = [];
+    let key = Object.keys(filterConfig);
+
+    if(key.length == 0){ 
+        for(let i=skip;i<top+skip;++i){
+            if(i > adList.length){
+                break;
+            }
+            adsToShow.push(adList[i]);
+        }
+    } else if(key[0] == 'vendors'){
+        let vendors = filterConfig[[key[0]]];
+        let skipped_ads = 0;
+        let i = 0;
+        while(skipped_ads < skip && i < adList.length){
+            if(vendors.includes(adList[i].vendor)){
+                skipped_ads++;
+            }
+            i++;
+        }
+        while(i < adList.length && adsToShow.length < top){
+            if(vendors.includes(adList[i].vendor)){
+                adsToShow.push(adList[i]);
+            }
+            i++;
+        }
+    }else if(key[0] == 'hashTags'){
+        let hashTags = filterConfig[[key[0]]];
+        let skipped_ads = 0;
+        let i = 0;
+        while(skipped_ads<skip && i < adList.length){
+            for(let j = 0;j<adList[i].hashTags.length;++j){
+                if(hashTags.includes(adList[i].hashTags[j])){
+                    skipped_ads++;
+                    break;
+                }
+            }
+            i++;
+        }
+        while(i < adList.length && adsToShow.length < top){
+            for(let j = 0;j<adList[i].hashTags.length;++j){
+                if(hashTags.includes(adList[i].hashTags[j])){
+                    adsToShow.push(adList[i]);
+                    break;
+                }
+            }
+            i++;
+        }
+    }
+    return adsToShow;
+});
+ let getAd = (function(id){
+    for (let i =0; i<adList.length;++i){
+        if(adList[i].id == id){
+            return adList[i];
+        }
+    }
+    return {};
+});
+
+let validateAd = (function(ad){
+    if (typeof(ad.id) == 'string' &&
+    typeof(ad.description) == 'string' && ad.description.length < 200  &&
+    ad.createdAt instanceof Date &&  
+    typeof(ad.link) == 'string' &&
+    typeof(ad.vendor) == 'string' && ad.vendor.length != 0 &&
+    ad.hashTags instanceof Array &&
+    typeof(ad.discount) == 'string' &&  ad.discount.length != 0 &&
+    ad.validUntil instanceof Date){
+        return true;
+    }else{
+        return false;
+    }
+});
+
+let addAd = (function(ad){
+    for (let i =0; i<adList.length;++i){
+        if(adList[i].id === ad.id){
+            return false;
+        }
+    }
+    if(validateAd(ad)){
+        adList.push(ad);
+        return true;
+    }else{
+        return false;
+    }
+});
+let editAd = (function(id,changes){
+    let keys = Object.keys(changes);
+    if(keys.includes('id') || keys.includes('author') || keys.includes('createdAt')){
+        return false;
+    }
+
+    let ind;
+    for(let i=0;i<=adList.length;++i){
+        if(i == adList.length){
+            return false;
+        }else if(adList[i].id === id){
+            ind = i;
+            break;
+        }
+    }
+    new_ad = adList[ind];
```

Comment:
Here you didn't copy object, you just copied it's link. So when you change keys of `new_ad` below you change it for array element as well. In order to avoid it you need to **copy** the array entry using `Object.assign` or destructuring (`let newAd = {...adList[ind]}`)
Second thing: please don't introduce variables without using `const`, `let` or `var`. And also please use camelCase as a Javascript variables notation.